# Lab4-Assignment about Named Entity Recognition and Classification

This notebook describes the assignment of Lab 4 of the text mining course. We assume you have succesfully completed Lab1, Lab2 and Lab3 as welll. Especially Lab2 is important for completing this assignment.

**Learning goals**
* going from linguistic input format to representing it in a feature space
* working with pretrained word embeddings
* train a supervised classifier (SVM)
* evaluate a supervised classifier (SVM)
* learn how to interpret the system output and the evaluation results
* be able to propose future improvements based on the observed results


## Credits
This notebook was originally created by [Marten Postma](https://martenpostma.github.io) and [Filip Ilievski](http://ilievski.nl) and adapted by Piek vossen

## [Points: 18] Exercise 1 (NERC): Training and evaluating an SVM using CoNLL-2003

**[4 point] a) Load the CoNLL-2003 training data using the *ConllCorpusReader* and create for both *train.txt* and *test.txt*:**

    [2 points]  -a list of dictionaries representing the features for each training instances, e..g,
    ```
    [
    {'words': 'EU', 'pos': 'NNP'}, 
    {'words': 'rejects', 'pos': 'VBZ'},
    ...
    ]
    ```

    [2 points] -the NERC labels associated with each training instance, e.g.,
    dictionaries, e.g.,
    ```
    [
    'B-ORG', 
    'O',
    ....
    ]
    ```

In [29]:
from nltk.corpus.reader import ConllCorpusReader

train = ConllCorpusReader(
    'CONLL2003',
    'train.txt',
    ['words', 'pos', 'ignore', 'chunk']
)

training_features = []
training_gold_labels = []

for token, pos, ne_label in train.iob_words():
    a_dict = {
        'words': token,
        'pos': pos
    }
    training_features.append(a_dict)
    training_gold_labels.append(ne_label)

In [30]:
test = ConllCorpusReader(
    'CONLL2003',
    'test.txt',
    ['words', 'pos', 'ignore', 'chunk']
)

test_features = []
test_gold_labels = []

for token, pos, ne_label in test.iob_words():
    a_dict = {
        'words': token,
        'pos': pos
    }
    test_features.append(a_dict)
    test_gold_labels.append(ne_label)

**[2 points] b) provide descriptive statistics about the training and test data:**
* How many instances are in train and test?
* Provide a frequency distribution of the NERC labels, i.e., how many times does each NERC label occur?
* Discuss to what extent the training and test data is balanced (equal amount of instances for each NERC label) and to what extent the training and test data differ?

Tip: you can use the following `Counter` functionality to generate frequency list of a list:

In [31]:
from collections import Counter 

my_list=[1,2,1,3,2,5]
Counter(my_list)

print("Train instances:", len(training_gold_labels))
print("Test instances:", len(test_gold_labels))

train_label_distribution = Counter(training_gold_labels)
test_label_distribution = Counter(test_gold_labels)

print("Train label distribution:")
print(train_label_distribution)

print("\nTest label distribution:")
print(test_label_distribution)

Train instances: 203621
Test instances: 46435
Train label distribution:
Counter({'O': 169578, 'B-LOC': 7140, 'B-PER': 6600, 'B-ORG': 6321, 'I-PER': 4528, 'I-ORG': 3704, 'B-MISC': 3438, 'I-LOC': 1157, 'I-MISC': 1155})

Test label distribution:
Counter({'O': 38323, 'B-LOC': 1668, 'B-ORG': 1661, 'B-PER': 1617, 'I-PER': 1156, 'I-ORG': 835, 'B-MISC': 702, 'I-LOC': 257, 'I-MISC': 216})


The training and test data show many more instances of 'O' than anything else, so the data is not "balanced". Approximately 80% of each set is composed of 'O'. Although, among the tokens receiving NERC labels, the data is relatively balanced between Loc, Per, and Org. As expected labels tagged with B outnumber labels tagged with I. Furthermore, the ratio of the different labels between the training and test data is similar. For all labels, the training instances are approximately 4-5 times more than the test instances. 

**[2 points] c) Concatenate the train and test features (the list of dictionaries) into one list. Load it using the *DictVectorizer*. Afterwards, split it back to training and test.**

Tip: You’ve concatenated train and test into one list and then you’ve applied the DictVectorizer.
The order of the rows is maintained. You can hence use an index (number of training instances) to split the_array back into train and test. Do NOT use: `
from sklearn.model_selection import train_test_split` here.


In [ ]:
from sklearn.feature_extraction import DictVectorizer

In [ ]:
vec = DictVectorizer()
the_array = # your code here

**[4 points] d) Train the SVM using the train features and labels and evaluate on the test data. Provide a classification report (sklearn.metrics.classification_report).**
The train (*lin_clf.fit*) might take a while. On my computer, it took 1min 53s, which is acceptable. Training models normally takes much longer. If it takes more than 5 minutes, you can use a subset for training. Describe the results:
* Which NERC labels does the classifier perform well on? Why do you think this is the case?
* Which NERC labels does the classifier perform poorly on? Why do you think this is the case?

In [8]:
from sklearn import svm

In [9]:
lin_clf = svm.LinearSVC()

In [3]:
##### [ YOUR CODE SHOULD GO HERE ]
# lin_clf.fit( # your code here

**[6 points] e) Train a model that uses the embeddings of these words as inputs. Test again on the same data as in 2d. Generate a classification report and compare the results with the classifier you built in 2d.**

In [10]:
# your code here

## [Points: 10] Exercise 2 (NERC): feature inspection using the [Annotated Corpus for Named Entity Recognition](https://www.kaggle.com/abhinavwalia95/entity-annotated-corpus)
**[6 points] a. Perform the same steps as in the previous exercise. Make sure you end up for both the training part (*df_train*) and the test part (*df_test*) with:**
* the features representation using **DictVectorizer**
* the NERC labels in a list

Please note that this is the same setup as in the previous exercise:
* load both train and test using:
    * list of dictionaries for features
    * list of NERC labels
* combine train and test features in a list and represent them using one hot encoding
* train using the training features and NERC labels

In [21]:
# Imports
import pandas as pd
from sklearn.feature_extraction import DictVectorizer

In [ ]:
# Load dataset
path = "ner_dataset.csv"

kaggle_dataset = pd.read_csv(
    path,
    encoding="latin1",
    on_bad_lines="skip"
)

Inspect dataset

In [9]:
len(kaggle_dataset)

1048575

In [10]:
kaggle_dataset.head()

,Sentence #,Word,POS,Tag
0,Sentence: 1,Thousands,NNS,O
1,NaN,of,IN,O
2,NaN,demonstrators,NNS,O
3,NaN,have,VBP,O
4,NaN,marched,VBN,O


In [11]:
kaggle_dataset.columns

Index(['Sentence #', 'Word', 'POS', 'Tag'], dtype='object')

In [ ]:
# Split dataset
df_train = kaggle_dataset[:100000]
df_test = kaggle_dataset[100000:120000]
print(len(df_train), len(df_test))

100000 20000


In [15]:
# Training features and labels
training_features = []
training_gold_labels = []

for _, row in df_train.iterrows():
    feature_dict = {
        "word": row["Word"],
        "pos": row["POS"]
    }

    training_features.append(feature_dict)
    training_gold_labels.append(row["Tag"])

In [16]:
# Test features and labels
test_features = []
test_gold_labels = []

for _, row in df_test.iterrows():
    feature_dict = {
        "word": row["Word"],
        "pos": row["POS"]
    }

    test_features.append(feature_dict)
    test_gold_labels.append(row["Tag"])

In [17]:
print(training_features[:5])
print(training_gold_labels[:5])

[{'word': 'Thousands', 'pos': 'NNS'}, {'word': 'of', 'pos': 'IN'}, {'word': 'demonstrators', 'pos': 'NNS'}, {'word': 'have', 'pos': 'VBP'}, {'word': 'marched', 'pos': 'VBN'}]
['O', 'O', 'O', 'O', 'O']


In [ ]:
# Vectorization using one-hot encoded features
vec = DictVectorizer()

all_features = training_features + test_features
feature_matrix = vec.fit_transform(all_features)

# Split back into training and test matrices
train_array = feature_matrix[:len(training_features)]
test_array = feature_matrix[len(training_features):]

In [20]:
print(train_array.shape)
print(test_array.shape)

(100000, 11897)
(20000, 11897)


**[4 points] b. Train and evaluate the model and provide the classification report:**
* use the SVM to predict NERC labels on the test data
* evaluate the performance of the SVM on the test data

Analyze the performance per NERC label.

In [22]:
# Imports
from sklearn import svm
from sklearn.metrics import classification_report

In [25]:
# Train SVM
linear_svm = svm.LinearSVC(max_iter=10000)

linear_svm.fit(train_array, training_gold_labels)

LinearSVC(max_iter=10000)

In [ ]:
predictions = linear_svm.predict(test_array)

# Classification Report
print(classification_report(test_gold_labels, predictions, zero_division=0))

              precision    recall  f1-score   support

       B-art       0.00      0.00      0.00         4
       B-eve       0.00      0.00      0.00         0
       B-geo       0.80      0.76      0.78       741
       B-gpe       0.96      0.92      0.94       296
       B-nat       1.00      0.50      0.67         8
       B-org       0.64      0.51      0.57       397
       B-per       0.81      0.53      0.64       333
       B-tim       0.91      0.76      0.83       393
       I-art       0.00      0.00      0.00         0
       I-eve       0.00      0.00      0.00         0
       I-geo       0.74      0.50      0.60       156
       I-gpe       1.00      0.50      0.67         2
       I-nat       0.80      1.00      0.89         4
       I-org       0.65      0.44      0.53       321
       I-per       0.42      0.90      0.57       319
       I-tim       0.41      0.08      0.14       108
           O       0.98      0.99      0.99     16918

    accuracy              

Performance analysis

The classifier performs best on the O label, with an F1-score of 0.99. This is expected, since most tokens in the dataset belong to this class. The model also performs well on labels such as B-gpe (0.94 F1), B-tim (0.83 F1), and B-geo (0.78 F1). I-nat also has a high score, although this is less meaningful statistically because there are only 4 examples in the test set.

Performance is noticeably lower for labels such as B-org (0.57 F1), I-org (0.53 F1), and especially I-tim (0.14 F1). For I-tim, the recall is very low, meaning the model often misses words that are part of longer time expressions. I-per shows a different issue: recall is high (0.90), but precision is much lower (0.42), meaning that the classifier predicts this label too often and makes many false positives.

The rarest labels, such as B-art and B-eve, receive scores of 0.00 because there are very few or no examples for the classifier to learn from. Overall, the results show that the classifier performs much better on common labels than on rare ones. Even though the overall accuracy is high at 0.94, the macro average F1-score of 0.52 shows that performance differs quite a lot across the individual NERC labels.

## End of this notebook